snowfall/training/graph.py

```diff
@@ -0,0 +1,26 @@
+from functools import lru_cache
+from typing import Iterable
+
+import k2
+from k2 import Fsa, SymbolTable
+
+
+class TrainingGraphCompiler:
+    def __init__(self, L: Fsa, vocab: SymbolTable, oov: str = '<UNK>'):
+        self.L = L
+        self.vocab = vocab
+        self.oov = oov
+
+    def compile(self, texts: Iterable[str]) -> Fsa:
+        decoding_graphs = k2.create_fsa_vec([self.compile_one_and_cache(text) for text in texts])
+        decoding_graphs.scores.requires_grad_(False)
+        return decoding_graphs
+
+    @lru_cache(maxsize=50000)
+    def compile_one_and_cache(self, text: str) -> Fsa:
```

Comment:
It may be more efficient to construct a list of linear fsas with a single call of `k2.linear_fsa`.
Not sure when the cache is in scope.

https://github.com/k2-fsa/k2/blob/66da100fc013cbd9a4d7f6ffaf6de3601c6e9ee8/k2/python/k2/fsa_algo.py#L17

```python
def linear_fsa(symbols: Union[List[int], List[List[int]]]) -> Fsa:
    '''Construct an linear FSA from symbols.
    Note:
      The scores of arcs in the returned FSA are all 0.
    Args:
      symbols:
        A list of integers or a list of list of integers.
    Returns:
      An FSA if the input is a list of integers.
      A vector of FSAs if the input is a list of list of integers.
    '''
```

Reply:
The discussion moved here: https://github.com/k2-fsa/k2/pull/476

It seems that since the intersection has to be done (?) on the CPU anyway, it's better to construct each graph on the CPU, and I'd prefer to do it individually for now (otherwise you can get a batch with a mix of cached and non-cached texts and the logic to handle that properly is a bit more complex). I can modify this later if it's still too slow.

The `lru_cache` caches the whole function conditioned on the input - if the argument (text) is present in the cache, it won't execute the function basically.

Reply:
Yeah, agree with @pzelasko here, considering we may shuffle between epoch, cache by text instead of `texts` woulb be much simpler.